### PR TITLE
feat(edit-page-v2): Unlocking a Locked Page Feature #28041

### LINKED ISSUE #28041: Edit Page V2: Unlocking a Locked Page Feature
### Parent
#27546 

### User Story
As a content editor, I want to be able to unlock a page that is locked by another user with proper permissions, so I can make necessary edits.

**PAGE CAN'T BE LOCKED FROM UVE (Universal Visual Editor)**

```[tasklist]
### Acceptance Criteria
- [ ] The user should see a notification bar indicating that the page is locked by another user.
- [ ] There should be a button available to unlock the page (if user has permission to unlock)
- [ ] The user should be able to unlock the page by clicking the button.
- [ ] Once unlocked, the user should have full editing capabilities on the page.
- [ ] While unlock request is happening we should see loading animation to the button
```

### Design

Page is locked and can be unlocked
image

Page is locked and CAN'T be unlocked
image

### Proposed Objective
Core Features

### Proposed Priority
Priority 3 - Average

### External Links... Slack Conversations, Support Tickets, Figma Designs, etc.:
- [Figma locked can unlock](https://www.figma.com/file/trbwjLuf5F6KaU1rFucNeg/Edit-Page?type=design&node-id=2901%3A71326&mode=design&t=9SmLxqID2f4Uk5ZB-1)
- [Figma locked can't unlock](https://www.figma.com/file/trbwjLuf5F6KaU1rFucNeg/Edit-Page?type=design&node-id=2901%3A71517&mode=design&t=9SmLxqID2f4Uk5ZB-1).
- [Prototype](https://www.figma.com/proto/trbwjLuf5F6KaU1rFucNeg/Edit-Page?page-id=1902%3A6704&type=design&node-id=2901-71326&viewport=1120%2C-7438%2C0.38&t=LhVcv54HwQAXbNJ7-1&scaling=scale-down&starting-point-node-id=2901%3A71326&mode=design)

### Assumptions & Initiation Needs
- There is a property in the page API `canUnlock`
- The locked page is not being actively edited by the user who locked it.

### Quality Assurance Notes & Workarounds
N/A

### PR DESCRIPTION
**Pull request by KevinDavilaDotCMS**
### Proposed Changes
* Now the user can unlock a locked page

User have a permission to unlock:

https://github.com/dotCMS/core/assets/144152756/940b0f81-0419-4917-a297-cf42783e96cb

User dont have a permission to unlock:

https://github.com/dotCMS/core/assets/144152756/6aab89ed-0b4e-4975-b284-b1a821fa57f9

### CODE REVIEW

**Review by zJaaal — CHANGES_REQUESTED** [review]

**Review by zJaaal — APPROVED** [review]

File: core-web/libs/portlets/edit-ema/portlet/src/lib/dot-ema-shell/store/dot-ema.store.spec.ts [hunk 1]
```diff
@@ -1000,6 +1089,15 @@ describe('EditEmaStore', () => {
...
             });
 
+            it('should call unlock page service', () => {
+                const dotContentletLockerService = spectator.inject(DotContentletLockerService);
+                const spyUnlock = jest.spyOn(dotContentletLockerService, 'unlock');
+
+                spectator.service.unlockPage('123');
+
+                expect(spyUnlock).toHaveBeenCalledWith('123');
+            });
```
Comment (fmontes) [inline core-web/libs/portlets/edit-ema/portlet/src/lib/dot-ema-shell/store/dot-ema.store.spec.ts:1099]: This updater patch the state and you're not checking that change.

File: core-web/libs/portlets/edit-ema/portlet/src/lib/dot-ema-shell/store/dot-ema.store.ts [hunk 2]
```diff
@@ -408,6 +427,35 @@ export class EditEmaStore extends ComponentStore<EditEmaState> {
...
+                                editorData: {
+                                    ...state.editorData,
+                                    page: {
+                                        ...state.editorData.page,
+                                        isLocked: false
+                                    }
+                                }
+                            }));
+
+                            this.updateEditorState(EDITOR_STATE.IDLE);
```
Comment (fmontes) [inline core-web/libs/portlets/edit-ema/portlet/src/lib/dot-ema-shell/store/dot-ema.store.ts:448]: Add this to the `patchState`.

File: core-web/libs/portlets/edit-ema/portlet/src/lib/dot-ema-shell/store/dot-ema.store.ts [hunk 3]
```diff
@@ -620,7 +668,19 @@ export class EditEmaStore extends ComponentStore<EditEmaState> {
...
+        isDefaultVariant,
+        canEditVariant,
+        isLocked
+    }: {
+        isDefaultVariant: boolean;
+        canEditVariant: boolean;
+        isLocked: boolean;
+    }): EDITOR_MODE {
+        if (isLocked) {
+            return EDITOR_MODE.LOCKED;
```
Comment (fmontes) [inline core-web/libs/portlets/edit-ema/portlet/src/lib/dot-ema-shell/store/dot-ema.store.ts:681]: 🔑😍

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.html [hunk 4]
```diff
@@ -54,7 +57,8 @@
                     !!es.contentletArea &&
                     !es.editorData.device &&
                     es.editor.page.canEdit &&
-                    (es.state === editorState.IDLE || es.state === editorState.DRAGGING)
+                    (es.state === editorState.IDLE || es.state === editorState.DRAGGING) &&
+                    !es.editorData.page.isLocked
```
Comment (fmontes) [inline core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.html:61]: This conditionals are out of hand... nothing to do at the moment tho.

File: core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts [hunk 5]
```diff
@@ -322,6 +322,10 @@ export class EditEmaEditorComponent implements OnInit, OnDestroy {
         return updatedRendered;
     }
 
+    unlockPage(inode: string) {
+        this.store.unlockPage(inode);
+    }
```
Comment (fmontes) [inline core-web/libs/portlets/edit-ema/portlet/src/lib/edit-ema-editor/edit-ema-editor.component.ts:327]: Do we need this here? aren't we doing this in the toolbar?

**Review by fmontes — APPROVED** [review]
Approve after the comments and suggestions.

File: core-web/libs/portlets/edit-ema/portlet/src/lib/dot-ema-shell/store/dot-ema.store.spec.ts (hunk 1 above)
Comment (KevinDavilaDotCMS) [reply core-web/libs/portlets/edit-ema/portlet/src/lib/dot-ema-shell/store/dot-ema.store.spec.ts:1099]: Fixed!

**Pull request merged**